Support INSERT/COPY/VACUUM on AO tables with unique indexes

soumyadeep2007 opened the pull request:
Please review commits individually.

huansong commented on src/backend/commands/indexcmds.c:
```diff
@@ -946,10 +946,23 @@ DefineIndex(Oid relationId,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("append-only tables do not support unique indexes")));
+
+		if (stmt->concurrent)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("append-only tables do not support unique indexes built concurrently")));
```
Shall we add a test case for this (CREATE INDEX CONCURRENTLY)?

soumyadeep2007 replied:
Its already banned at a higher level (even for heap) in gram.y, and we already have test coverage for it:

```
if (n->concurrent)
						ereport(ERROR,
								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
								 errmsg("CREATE INDEX CONCURRENTLY is not supported")));
```

I just added one more conditional here to ensure that anyone enabling it for all will know that there are special restrictions for AO/AOCO tables, which would need additional work.

huansong commented on src/backend/access/appendonly/appendonlyblockdirectory.c:
```diff
@@ -1172,11 +1305,100 @@ write_minipage(AppendOnlyBlockDirectory *blockDirectory,
+	 * is created in memory, in a subsequent call to insert_new_entry(),
+	 * followed by a write_minipage() which will make this "update" persistent.
+	 */
+	minipagePerColumnGroup->numMinipageEntries--;
```
Don't have a strong opinion but this doesn't actually clear the memory, is it? Should we just call `clear_minipage` since we could only have one minipage entry at this point?

soumyadeep2007 replied:
Actually, we could have multiple entries at this point. Imagine a segfile already had 5 minipage entries worth of rows and we are inserting a new row. Then the placeholder entry will be the 6th entry in the minipage.

huansong replied:
Hmm I see. So the existing minipage entries will be loaded when initializing the in-memory block directory content (AppendOnlyBlockDirectory_Init_forInsert->load_last_minipage). Didn't notice that before. Thanks!

huansong commented on src/backend/access/appendonly/appendonlyam_handler.c:
```diff
@@ -499,8 +517,29 @@ appendonly_index_fetch_tuple(struct IndexFetchTableData *scan,
-	appendonly_fetch(aoscan->aofetch, (AOTupleId *) tid, slot);
+	/*
+	 * No need to fetch the tuple if we are performing an existence check for
+	 * the tuple (we actually can't reliably as we might encounter the
```
`can't reliably *fetch/do so`

huansong reviewed (commented):
Great work! Had mostly just minor comments. Will play with the tests a bit more tomorrow.

soumyadeep2007 commented on src/test/isolation2/sql/ao_unique_index.sql:
```diff
@@ -0,0 +1,275 @@
+-- Tests to ensure that unique indexes work as expected w/ ao_row tables.
```
While that can be done for the other tests (vac, delete and update which are upcoming), tests for INSERT need the separation because the number of items in a varblock are different (2048 for AO vs 1818 for AOCO) which makes the number of rows to be inserted (to test minipage boundaries) different too.

soumyadeep2007 commented:
> Great work! Had mostly just minor comments. Will play with the tests a bit more tomorrow.

Thanks for the review! I am addressing the feedback and will be mostly done tomorrow.

A couple of queries that can help you play with the tests - helps get full picture of block directory entries (and their expanded minipages):

```
-- blkdir rows at high level
select blkdirrelid::regclass from pg_appendonly where relid = 'foo'::regclass;
select ctid, xmin, xmax, txid_status(xmin::text::bigint) as xmin_status, txid_status(xmax::text::bigint) as xmax_status, segno, columngroup_no, first_row_no from pg_aoseg.pg_aoblkdir_106518 order by 1,2::text,4,5,6;

-- expanded view of each minipage for each block directory row
select * from gp_toolkit.__gp_aoblkdir('foo'::regclass) order by 1,2,3,4,5;
```

soumyadeep2007 commented on src/backend/access/appendonly/appendonlyblockdirectory.c:
```diff
@@ -932,42 +1101,6 @@ extract_minipage(AppendOnlyBlockDirectory *blockDirectory,
 
 	ItemPointerCopy(&tuple->t_self, &minipageInfo->tupleTid);
-
-	/*
-	 * When crashes during inserts, or cancellation during inserts, there are
-	 * out-of-date minipage entries in the block directory. We reset those
-	 * entries here.
-	 */
-	Assert(fsInfo != NULL);
-	if (!blockDirectory->isAOCol)
-		eof = fsInfo->eof;
-	else
-		eof = ((AOCSFileSegInfo *) fsInfo)->vpinfo.entry[columnGroupNo].eof;
-
-	start = 0;
-	end = minipageInfo->numMinipageEntries - 1;
-	while (start <= end)
-	{
-		mid = (end - start + 1) / 2 + start;
-		if (minipageInfo->minipage->entry[mid].fileOffset > eof)
-			end = mid - 1;
-		else if (minipageInfo->minipage->entry[mid].fileOffset < eof)
-			start = mid + 1;
-		else
-		{
-			found = true;
-			break;
-		}
-	}
-
-	minipageInfo->numMinipageEntries = 0;
-	if (found)
-		minipageInfo->numMinipageEntries = mid;
-	else if (start > 0)
-	{
-		minipageInfo->numMinipageEntries = start;
-		Assert(minipageInfo->minipage->entry[start - 1].fileOffset < eof);
-	}
```
We trip the assertion failure when doing a unique index check (during an insert) because `blockDirectory->currentSegmentFileInfo` is not set up (it is only set up when we need to read a segfile, which we never need to do for unique index checks. See `AppendOnlyBlockDirectory_GetEntry()`)

If you add back the code in this function and run the following it is easy to repro:
```sql
CREATE TABLE unique_index_ao_row (a INT unique) USING ao_row DISTRIBUTED REPLICATED;
CREATE
INSERT INTO unique_index_ao_row SELECT * FROM generate_series(1, 329729);
INSERT 329729
-- should conflict
INSERT INTO unique_index_ao_row VALUES (1);
ERROR:  Unexpected internal error (assert.c:44)  (seg0 192.168.0.148:7002 pid=2667731) (assert.c:44)
DETAIL:  FailedAssertion("!(fsInfo != ((void *)0))", File: "appendonlyblockdirectory.c", Line: 1117)
```

Now, we could be conservative and keep this code around for non-unique-index-check code paths. I think I removed this entirely for all code paths after I discussed w/ @ashwinstar . Question we asked is: "Why is fixing these entries even necessary for cancels/crashes?" By MVCC rules on the block dir relation, incorrect entries from cancelled/aborted/crashed inserts would simply be invisible as their corresponding blkdir row(s) would be invisible. 

Another data point that led us to believe that this code is vestigial is that none of the tests crashed, even those dealing with AO crashes/cancels/aborts.

I'll highlight this removal in the commit message for now and keep this removal, unless anyone feels otherwise.

ashwinstar replied:
As discussed before I do continue to vote to remove this code. Though recommend having it as a separate commit and not bundled with the new functionality we are trying to add, unless really have strong reasoning for the same.

soumyadeep2007 replied:
Done.

huansong commented on src/test/isolation2/sql/aocs_unique_index.sql:
```diff
@@ -0,0 +1,275 @@
+-- Case 5: Conflict with to-be-aborted transaction------------------------------
+--
+-- 1. Uncommitted tx 1 has inserted non-conflicting key = 0.
+-- 2. Uncommitted tx 2 has inserted (161 * 4090 + 2 = 658492 rows), which spans
```
nit: Tx 2 actually inserts 658491 instead of 658492 rows.

huansong commented on src/backend/access/appendonly/appendonlyblockdirectory.c:
```diff
@@ -1172,11 +1305,100 @@ write_minipage(AppendOnlyBlockDirectory *blockDirectory,
+ * 	(i) the current in-memory minipage is full
+ * 	(ii) at end of command.
+ *
+ * So we insert a placeholder row in the block directory table before the first
```
From my updated understanding, there could be existing entries in the current in-memory minipage content. And we will insert one more placeholder entry to that content. 
So, we should say we insert a placeholder *entry* instead of a placeholder *row*, right? I know that we explained later that:
```
Such a row whose last entry is a placeholder entry is called a placeholder row.
```
But still we are only inserting an entry. Whatever the existing minipage content has it is not "placeholder" per se. It's not a big issue but could confuse people IMO (at least I was a little bit initially :) ).

soumyadeep2007 replied:
Yes it is confusing. Fixed, thanks for pointing out!

huansong reviewed (commented):
More test related questions for my naivety.

soumyadeep2007 commented on src/test/isolation2/sql/ao_unique_index.sql:
```diff
@@ -0,0 +1,275 @@
+-- Case 4: Conflict with to-be-committed transaction----------------------------
+--
+-- 1. Uncommitted tx 1 has inserted non-conflicting key = 0.
+-- 2. Uncommitted tx 2 has inserted (161 * 2048 + 1 = 329729 rows), which spans
+--    2 block directory rows (1st row: [1,329728] ; 2nd row: [329729,329729])
```
NUM_MINIPAGE_ENTRIES = 161, the max number of minipage entries in a minipage.

soumyadeep2007 commented on src/test/isolation2/sql/ao_unique_index.sql:
```diff
@@ -0,0 +1,275 @@
+----------------- More concurrent tests with fault injection ------------------
+--------------------------------------------------------------------------------
+
+-- Case 7: Conflict with to-be-committed transaction while only 1 placeholder
+-- row exists in the block directory--------------------------------------------
+--
+-- This case highlights the importance of the placeholder row, inserted at the
+-- beginning of an INSERT command.
+--
+-- 1. Uncommitted Tx 1 has inserted 3 out of its 10 rows and is suspended.
+-- 2. Tx 2 inserts a conflicting row and blocks on Tx 1.
+-- 3. Tx 3 inserts a non-conflicting row within the range [4,10] and is
+--    immediately successful. (Index entries have been written only for [1,3] so
+--    far, so conflicts shouldn't arise)
+-- 4. Tx 4 inserts a non-conflicting row in range [11, ..) and should be
+--    immediately successful.
+-- 5. Now Tx 1 resumes and tries to insert a row in range [4,10] and reports a
+--    unique constraint violation with Tx 3.
+-- 6. Tx 2 succeeds as Tx 1 aborted.
```
The way I modeled these tests is that case 7 shows that the placeholder entry mechanism is necessary and case 8 shows that only 1 such placeholder entry is necessary to be inserted at the beginning of each insert (and not one per minipage insert aka block directory row insert)

These tests are very white-box and that's why I go into so much detail about the placeholder entry etc.

soumyadeep2007 replied:
> It seems that the Case 8 alone can showcase the necessity of placeholder row: let the unique index check logic realize that there's corresponding minipage in memory that it can check even the content isn't on disk. Lmk what I understand incorrectly.

The unique index check code can't realize that something is in memory and not on disk: that is why we create the placeholder row on disk, so that the check code can read the placeholder entry from the table (via forced sysscan) and figure out that there is a conflict or not.

Yes as mentioned case 8 is a generalization of case 7. But I want to keep both as 7 is more minimal and easier to understand and is like a warm-up for case 8.

soumyadeep2007 commented:
Update:

1. Rebased against latest master
2. Modified vacuum commit to formally add full support for vacuum - added coverage.
3. Addressed @huansong 's feedback

jimmyyih commented on src/include/cdb/cdbaocsam.h:
```diff
@@ -318,12 +318,6 @@ extern bool aocs_fetch(AOCSFetchDesc aocsFetchDesc,
 					   TupleTableSlot *slot);
 extern void aocs_fetch_finish(AOCSFetchDesc aocsFetchDesc);
-
-extern AOCSUpdateDesc aocs_update_init(Relation rel, int segno);
```
With these unused functions being removed, we can get rid of `AOCSUpdateDesc` and `AOCSUpdateDescData` too.

soumyadeep2007 replied:
Fair point! Will do!

jimmyyih commented on src/test/isolation2/expected/ao_unique_index_vacuum.out:
```diff
@@ -0,0 +1,21 @@
+SET gp_appendonly_enable_unique_index TO ON;
```
I don't see a corresponding sql file for this expected output file.  Is this an old file that was replaced by the autogenerating `uao/ao_unique_index_vacuum.source` test?

soumyadeep2007 replied:
Yep, forgot to hit rm on those. Will fix

jimmyyih commented on src/test/isolation2/expected/aocs_unique_index_vacuum.out:
```diff
@@ -0,0 +1,21 @@
+SET gp_appendonly_enable_unique_index TO ON;
```
I don't see a corresponding sql file for this expected output file.  Is this an old file that was replaced by the autogenerating `uao/ao_unique_index_vacuum.source` test?

soumyadeep2007 replied:
Yep, forgot to hit rm on those. Will fix

jimmyyih commented on src/backend/executor/execIndexing.c:
```diff
@@ -264,6 +264,10 @@ ExecCloseIndices(ResultRelInfo *resultRelInfo)
  *		If 'arbiterIndexes' is nonempty, noDupErr applies only to
  *		those indexes.  NIL means noDupErr applies to all indexes.
  *
+ *		GPDB: gp_bypass_unique_check is introduced so that routines
+ *		such as AO vacuum which don't need to run uniqueness checks
+ *		while inserting tuples can do so.
+ *
```
Is this a temporary thing or will it be removed when DELETE/UPDATE support is introduced?

soumyadeep2007 replied:
Its permanent. The changes to DELETE/UPDATE won't affect this.

soumyadeep2007 replied:
In case you wanted a preview for reference, these are the 2 commits for DELETE/UPDATE: https://github.com/soumyadeep2007/gpdb/tree/ao_uniq_delete

jimmyyih commented:
From one of your commit messages:
> We also ban unique index creation if the AO table has a relation version
<= AORelationVersion_PG83 and if it already has a block directory.

Should the "and if it already has a block directory" part be removed?  The logic in `ValidateRelationVersionForUniqueIndex` only checks for formatversion less than AORelationVersion_PG12.  If I had an AO table with formatversion AORelationVersion_PG83 and no block directory, it would error out.  If I had an AO table with formatversion AORelationVersion_PG12 and an existing block directory, creating a unique index still works.

soumyadeep2007 commented:
> From one of your commit messages:
> 
> > We also ban unique index creation if the AO table has a relation version
> > <= AORelationVersion_PG83 and if it already has a block directory.
> 
> Should the "and if it already has a block directory" part be removed? The logic in `ValidateRelationVersionForUniqueIndex` only checks for formatversion less than AORelationVersion_PG12. If I had an AO table with formatversion AORelationVersion_PG83 and no block directory, it would error out. If I had an AO table with formatversion AORelationVersion_PG12 and an existing block directory, creating a unique index still works.

Yes but the routine is called conditionally:

```c
/* Additional version checks needed if block directory already exists */
if (OidIsValid(blkdirrelid))
	ValidateRelationVersionForUniqueIndex(rel);
```

I'm adding a comment to make that explicit.

huansong reviewed (approved):
LGTM!

ashwinstar commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -681,6 +681,15 @@ aoco_index_fetch_tuple(struct IndexFetchTableData *scan,
+	 * transactions. Failure to set this will lead to use of a garbage value
+	 * in certain code, such as that for unique index checks.
+	 */
+	if (all_dead)
```
any specific reason to have this initialized kind of in the middle of the function? Seems better located at the start instead

soumyadeep2007 replied:
Fair, moving it to the end as these params are peripheral to the intent of the function. Also typically output params signifying side effects are typically populated at the end.

ashwinstar commented:
Thank you so much for capturing the design details at length in the commit message. Let's help our future selves by updating this [Append Optimized Design README](https://github.com/greenplum-db/gpdb/blob/master/src/backend/access/appendonly/README.md) to reflect the design of unique indexes.

ashwinstar commented on src/backend/access/aocs/aocsam.c:
```diff
@@ -2150,3 +2149,30 @@ aocs_addcol_emptyvpe(Relation rel,
 }
+
+/*
+ * Lighter version of aocs_fetch(), where we merely check for the existence
```
Should be a comment about which snapshot is used to perform this checking?

soumyadeep2007 replied:
This is made explicit with the asserts in the `index_fetch_tuple_exists` implementations.

ashwinstar commented on src/backend/access/aocs/aocsam.c:
```diff
@@ -2150,3 +2149,30 @@ aocs_addcol_emptyvpe(Relation rel,
+	Assert(segmentFileNum != InvalidFileSegNumber);
+	Assert(rowNum != InvalidAORowNum);
+
+	if (aocsFetchDesc->lastSequence[segmentFileNum] == InvalidAORowNum)
```
Interesting, not the change introduced by this PR, though this hints that even during the fetch (read) operation, `gp_fastsequence` table is read currently. Curious to learn why?

soumyadeep2007 replied:
Let me point you to 2 things:

Definition of lastSequence sparse array
```
/*
 * Array containing the maximum row number in each aoseg (to be consulted
 * during fetch). This is a sparse array as not all segments are involved
 * in a scan. Sparse entries are marked with InvalidAORowNum.
 *
 * Note:
 * If we have no updates and deletes, the total_tupcount is equal to the
 * maximum row number. But after some updates and deletes, the maximum row
 * number is always much bigger than total_tupcount, so this carries the
 * last sequence from gp_fastsequence.
 */
int64			lastSequence[AOTupleId_MultiplierSegmentFileNum];
```

and its use in appendonly_fetch() for optimizing BRIN:

```c
/*
 * This is an improvement for brin. BRIN index stores ranges of TIDs in
 * terms of block numbers and not specific TIDs, so it's possible that the
 * fetch function is called with a non-existent TID. The function
 * appendonly_fetch will access the block directory table first and cache
 * some MinipageEntrys. If we try to access the non-existent tid, a cache
 * miss will occur. And we need to search the btree on block directory
 * table. This is a vary slow operation. So a fast return path was added
 * here. If the rowNum is bigger than lastsequence, skip it.
 */
if (rowNum > aoFetchDesc->lastSequence[segmentFileNum])
{
	if (slot != NULL)
	{
		ExecClearTuple(slot);
	}
	return false;	/* row has been deleted or updated. */
}
```

Seems like this field was introduced solely for BRIN indexes. Since unique indexes can only be btree, this may/may not make sense. Then again the conditional doesn't hurt but is distracting. We could remove it for this branch.

We can also go the distance and only initialize `lastSequence` by reading `gp_fastsequence` iff we are fetching using a BRIN index (only trouble is that knowing if it is BRIN or not isn't easy in the fetch code)

ashwinstar replied:
I am trying to see if there is any downside to this lastsequence checking apart from the performance. Like we are not even providing flexibility on which snapshot is used for reading the fastsequence entry. I understand we are coding a special function for performing the unique checking, if not and reused the fetch function, then we don't wish this lastsequence logic to kick-in as wish to fetch blockdirectory for concurrent inserts as well. If this logic kicks-in then would return early based on lastsequence and ignore concurrent inserted index entries.

Yes hard to find use cases where this helps for indexes other than BRIN. Maybe lossy Bitmap index scans might help as those behave similarly to BRIN. So, seems worth tracking this as a separate item and tracing down the details.

ashwinstar commented on src/backend/access/aocs/aocsam.c:
```diff
@@ -2150,3 +2149,30 @@ aocs_addcol_emptyvpe(Relation rel,
+	if (aocsFetchDesc->lastSequence[segmentFileNum] == InvalidAORowNum)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("Row No. %ld in segment file No. %d is out of scanning scope for target relfilenode %u.",
```
`relfilenode` is very hard to consume as a user in error messages, please can you provide the relation name as well? Also, the error message must start in lowercase always.

soumyadeep2007 replied:
No longer having this ERROR message.

ashwinstar commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -302,10 +302,49 @@ get_insert_descriptor(const Relation relation)
+		insertDesc = aocs_insert_init(relation,
 											 ChooseSegnoForWrite(relation));
+		/*
+		 * If we have a unique index, insert a placeholder block directory row to
```
Hmm is this the only best place for inserting the place-holder row in the block directory? Mainly the function name being `get_insert_descriptor()` rarely provides an indication that will do any write or state change operation to the system.

soumyadeep2007 replied:
Yeah, the other alternative, which is to put it inside the `aocs_insert_init` function is bad as we don't want vacuum to be doing this, which is also a user. Let me think on this some more.

PS: I have renamed it to `get_or_create_insert_descriptor` in https://github.com/greenplum-db/gpdb/pull/13925. But even that isn't good enough.

ashwinstar commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -302,10 +302,49 @@ get_insert_descriptor(const Relation relation)
+					break;
+				}
+			}
+			Assert(firstNonDroppedColumn != -1);
```
not sure if we already ban 0-column CO tables or not. If not then we should do that as not sure the current code even works, plus this assertion will trip with it.

soumyadeep2007 replied:
One can't have a unique index without a column, so this is okay.

ashwinstar commented on src/include/access/tableam.h:
```diff
@@ -317,12 +317,18 @@ typedef struct TableAmRoutine
 									  Snapshot snapshot,
 									  TupleTableSlot *slot,
-									  bool *call_again, bool *all_dead);
+									  bool *call_again, bool *all_dead,
```
similar to `all_dead` seems we are missing initializing `call_again` as well, lets add code in AO for it

soumyadeep2007 replied:
Done!

ashwinstar commented on src/backend/access/table/tableam.c:
```diff
@@ -207,7 +207,7 @@ table_index_fetch_tuple_check(Relation rel,
 	slot = table_slot_create(rel, NULL);
 	scan = table_index_fetch_begin(rel);
 	found = table_index_fetch_tuple(scan, tid, snapshot, slot, &call_again,
```
Comment above states that tuple by tuple performing beginscan...fetchtuple... endscan mostly for heap doesn't matter. Does that reality hold true that optimization is not required here for AO and especially CO tables where I think during beginscan or init we allocate memory, define projection list, and such kind of things?

soumyadeep2007 replied:
You are right. We need to think on how to optimize this for AO/CO.

soumyadeep2007 replied:
### Performance analysis:

During my performance testing I found that with the current branch, for the worst case (live index tuple points to dead AO tuple), we take a substantial performance hit due to this aspect:

```sql
test=# create table foo(i int unique) using ao_row;
test=# begin; insert into foo select generate_series(1, 100000); abort;
test=# insert into foo select generate_series(1, 100000);
INSERT 0 100000
Time: 22694.428 ms (00:22.694)   <---------------------- BAD!
``` 

### Solution:

* So, I decided to stop obeying this begin..fetch..end business for every tuple. 
* I created a `AppendOnlyUniqueCheckDesc` in `AppendOnlyDMLState`, which contains an AppendOnlyBlockDirectory structure (which I initialize at first lookup) which I use for the lookups, using a `get_or_create_unique_check_desc()`, similar to `get_or_create_insert_desc()`.
* In the future this will also contain a reference to a visimap structure, which can be used to entertain DELETEs/UPDATEs.

Results are very promising:
```sql
test=# create table foo(i int unique) using ao_row;
CREATE TABLE
test=# begin; insert into foo select generate_series(1, 100000); abort;
test=# \timing
Timing is on.
test=# insert into foo select generate_series(1, 100000);
INSERT 0 100000
Time: 761.793 ms
```

For reference, heap performance:
```sql
test=# create table heap(i int unique);
CREATE TABLE
begin; insert into heap': begin; insert into foo select generate_series(1, 100000); abort;
test=# begin; insert into heap select generate_series(1, 100000); abort;
\timing on
test=# insert into heap select generate_series(1, 100000);
INSERT 0 100000
Time: 553.628 ms
```

#### 2 implementation choices:
* Make `appendonly_fetch_init()` and `appendonly_fetch_finish()` a no-op when `existence_check` is passed to these.
    * Pros:
           * No changes to tableam.c - it is still oblivious to AO vs heap
           * Guards against change to tableam code
    * Cons:
            * Per tuple function call overhead
            * The tableam APIs need to change (existence_check param) + non-AO callsites need to change etc.
            * We are kind of lying: we are not using the scan desc etc
* In `table_index_fetch_tuple_check()`, check if `RelationIsAppendOptimized()`, and if so call the corresponding `fetch_exists()` variant and return. For heap, perform existing begin..fetch..end stuff.
      * Pros:
           * Most performant
           * Most understandable, no lying and no forcing squares into round holes
           * No changes to table AM APIs: We can undo all of the `existence_check` changes.
      * Cons:
            * tableam.c has ugly dependency on AO
            * Does not guard against change to tableam code

I have a dirty patch for approach 2. I will discuss offline with @ashwinstar  and push the changes accordingly.

soumyadeep2007 replied:
Upon discussing w/ @ashwinstar, we are going w/ secret option no 3. We will be adding a new table AM call for performing an existence check and will be calling that instead for AO/CO tables from within table_index_fetch_tuple_check. For heap, we will leave the API unimplemented. And will also add comments everywhere.

ashwinstar commented on src/include/access/tableam.h:
```diff
@@ -317,12 +317,18 @@ typedef struct TableAmRoutine
 	 * that tuple. Index AMs can use that do avoid returning that tid in
 	 * future searches.
+	 *
+	 * GPDB: existence_check is an argument introduced to support existence
```
seems `existence_check` belongs more to `index_fetch_begin()` as a one-time setting when initializing the scan descriptor. With the change to `index_fetch_tuple()` is passed tuple by tuple on every call which seems unnecessary stack overhead as will be true/false for the life of the scan object.

soumyadeep2007 replied:
This is no longer necessary due to the fresh new table AM API.

ashwinstar commented on src/include/executor/executor.h:
```diff
@@ -618,7 +618,8 @@ extern TupleTableSlot *ExecGetReturningSlot(EState *estate, ResultRelInfo *relIn
 extern void ExecOpenIndices(ResultRelInfo *resultRelInfo, bool speculative);
 extern void ExecCloseIndices(ResultRelInfo *resultRelInfo);
 extern List *ExecInsertIndexTuples(TupleTableSlot *slot, EState *estate, bool noDupErr,
```
I really feel instead of messing with this API and changing the calls at so many places, this flag better belongs to Estate. When creating the estate in `AOCSSegmentFileFullCompaction()` can assign the true value and by default can have false. This way the changes get minimized with no API change as this is an exposed API and change will have an impact as PG extensions would need code modifications to compile against GPDB.

soumyadeep2007 replied:
Sounds great! Will do!

ashwinstar commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -847,6 +848,17 @@ struct config_bool ConfigureNamesBool_gp[] =
 
+	{
+		{"gp_appendonly_enable_unique_index", PGC_USERSET, DEVELOPER_OPTIONS,
+		 gettext_noop("Verify the append-only block checksum when reading."),
```
GUC text needs to be updated

soumyadeep2007 replied:
Done.

ashwinstar commented on src/backend/access/appendonly/appendonlyblockdirectory.c:
```diff
@@ -667,6 +674,168 @@ AppendOnlyBlockDirectory_GetEntry(
+ * Currently used to satisfy unique constraint checks.
+ */
+bool
+entry_exists(AppendOnlyBlockDirectory *blockDirectory,
```
`entry_exists()` is really generic name, understand its static function still maybe helpful if it at least has some block directory aspect in its name, no mandatory to address this comment.

soumyadeep2007 replied:
Renamed to `blkdir_entry_exists()`

ashwinstar commented on src/backend/access/appendonly/appendonlyblockdirectory.c:
```diff
@@ -667,6 +674,168 @@ AppendOnlyBlockDirectory_GetEntry(
+					  "(%d, %d, " INT64_FORMAT ")",
+				  0, segmentFileNum, rowNum)));
+
+	heapTupleDesc = RelationGetDescr(blkdirRel);
```
instead of heapTupleDesc would be helpful to call it blkdirTupleDesc

soumyadeep2007 replied:
Done.

ashwinstar commented:
I have completed the review for the INSERT/COPY commit. Will take up VACUUM-related code changes in the coming days.

soumyadeep2007 commented:
* Rebased against latest master
* Addressed all outstanding feedback
* Performance bug identified which has led to a slight redesign around the tableam interface: [Details](https://github.com/greenplum-db/gpdb/pull/14097#discussion_r990552086)
Please see the updated commit message for more context.
To summarize the changes, we have a fresh new table AM API function `index_fetch_tuple_exists`, implemented by AO/CO. This avoids the overhead of setting up and destroying scan descriptors in the beginscan..fetch..endscan construct in `table_index_fetch_tuple_check`
